File: lld/problems/snake-and-ladder/go/snakeandladder.go
// Package snakeandladder implements the classic Snake and Ladder LLD problem:
// a configurable board with snakes and ladders, an injectable dice (Strategy
// pattern) for deterministic testing, and a turn-based game engine that
// enforces the exact-landing-to-win rule.
package snakeandladder

import (
	"errors"
	"fmt"
	"math/rand"
)

// Dice rolls a value each turn. Implementations can be random or scripted.
type Dice interface {
	Roll() int
}

// StandardDice is a fair single six-sided die, seedable for reproducible runs.
type StandardDice struct {
	rng *rand.Rand
}

func NewStandardDice(seed int64) *StandardDice {
	return &StandardDice{rng: rand.New(rand.NewSource(seed))}
}

func (d *StandardDice) Roll() int {
	return d.rng.Intn(6) + 1
}

// ScriptedDice replays a fixed sequence of rolls, cycling once exhausted.
// This is the key to writing deterministic, non-flaky tests for the engine.
type ScriptedDice struct {
	Rolls []int
	pos   int
}

func NewScriptedDice(rolls ...int) *ScriptedDice {
	return &ScriptedDice{Rolls: rolls}
}

func (d *ScriptedDice) Roll() int {
	if len(d.Rolls) == 0 {
		return 1
	}
	v := d.Rolls[d.pos%len(d.Rolls)]
	d.pos++
	return v
}

// Entity is a board hazard/shortcut: a snake (head->tail) or ladder (bottom->top).
type Entity struct {
	Start int
	End   int
}

// Board holds the size and the snake/ladder map keyed by the cell a player lands on.
type Board struct {
	Size     int
	entities map[int]int
}

// NewBoard builds a board of the given size (last cell = size) with the supplied
// snakes and ladders. Entities with overlapping start/end are rejected to avoid
// ambiguous or infinite-loop configurations.
func NewBoard(size int, snakes, ladders []Entity) (*Board, error) {
	b := &Board{Size: size, entities: make(map[int]int)}
	for _, s := range snakes {
		if s.Start <= s.End {
			return nil, fmt.Errorf("snake start %d must be greater than end %d", s.Start, s.End)
		}
		if err := b.addEntity(s.Start, s.End); err != nil {
			return nil, err
		}
	}
	for _, l := range ladders {
		if l.Start >= l.End {
			return nil, fmt.Errorf("ladder start %d must be less than end %d", l.Start, l.End)
		}
		if err := b.addEntity(l.Start, l.End); err != nil {
			return nil, err
		}
	}
	return b, nil
}

func (b *Board) addEntity(from, to int) error {
	if from <= 1 || from >= b.Size {
		return fmt.Errorf("entity start %d must be within (1, %d)", from, b.Size)
	}
	if _, exists := b.entities[from]; exists {
		return fmt.Errorf("cell %d already has a snake or ladder", from)
	}
	b.entities[from] = to
	return nil
}

// resolve applies any snake/ladder at the landed-on cell, returning the final resting cell.
func (b *Board) resolve(cell int) int {
	if dest, ok := b.entities[cell]; ok {
		return dest
	}
	return cell
}

type Player struct {
	Name     string
	Position int
}

var (
	ErrNotEnoughPlayers = errors.New("need at least two players")
	ErrGameAlreadyOver  = errors.New("game already has a winner")
)

// Game drives turn order, dice rolls, and win detection for a single match.
type Game struct {
	Board   *Board
	Dice    Dice
	Players []*Player
	turn    int
	Winner  *Player
}

func NewGame(board *Board, dice Dice, playerNames []string) (*Game, error) {
	if len(playerNames) < 2 {
		return nil, ErrNotEnoughPlayers
	}
	players := make([]*Player, len(playerNames))
	for i, name := range playerNames {
		players[i] = &Player{Name: name, Position: 0}
	}
	return &Game{Board: board, Dice: dice, Players: players}, nil
}

// PlayTurn rolls the dice for the current player, applies the exact-landing rule
// and any snake/ladder at the destination, then rotates turn order to the next
// player. It returns the player who just moved.
func (g *Game) PlayTurn() (*Player, error) {
	if g.Winner != nil {
		return nil, ErrGameAlreadyOver
	}

	player := g.Players[g.turn]
	roll := g.Dice.Roll()
	target := player.Position + roll

	// Overshooting the last cell is not a legal move: the player stays put.
	if target <= g.Board.Size {
		player.Position = g.Board.resolve(target)
	}

	if player.Position == g.Board.Size {
		g.Winner = player
	} else {
		g.turn = (g.turn + 1) % len(g.Players)
	}
	return player, nil
}

// CurrentPlayer returns whose turn it is next.
func (g *Game) CurrentPlayer() *Player {
	return g.Players[g.turn]
}

// Play runs turns until a winner emerges, guarding against runaway loops.
func (g *Game) Play(maxTurns int) (*Player, error) {
	for i := 0; i < maxTurns; i++ {
		if _, err := g.PlayTurn(); err != nil {
			return nil, err
		}
		if g.Winner != nil {
			return g.Winner, nil
		}
	}
	return nil, fmt.Errorf("no winner after %d turns", maxTurns)
}

File: lld/problems/snake-and-ladder/python/snake_and_ladder.py
"""Snake and Ladder LLD — Python reference implementation.

A configurable board with snakes and ladders, an injectable dice (strategy
pattern) for deterministic testing, and a turn-based game engine that
enforces the exact-landing-to-win rule.
See ../go/snakeandladder.go for the original design writeup.
"""
from __future__ import annotations

import random
from dataclasses import dataclass
from typing import Dict, List, Optional, Protocol, Sequence


class Dice(Protocol):
    """Rolls a value each turn. Implementations can be random or scripted."""

    def roll(self) -> int:
        ...


class StandardDice:
    """A fair single six-sided die, seedable for reproducible runs."""

    def __init__(self, seed: Optional[int] = None):
        self._rng = random.Random(seed)

    def roll(self) -> int:
        return self._rng.randint(1, 6)


class ScriptedDice:
    """Replays a fixed sequence of rolls, cycling once exhausted.

    This is the key to writing deterministic, non-flaky tests for the
    engine.
    """

    def __init__(self, *rolls: int):
        self.rolls = list(rolls)
        self._pos = 0

    def roll(self) -> int:
        if not self.rolls:
            return 1
        v = self.rolls[self._pos % len(self.rolls)]
        self._pos += 1
        return v


@dataclass(frozen=True)
class Entity:
    """A board hazard/shortcut: a snake (head->tail) or ladder (bottom->top)."""

    start: int
    end: int


class BoardConfigError(Exception):
    pass


class Board:
    """Holds the size and the snake/ladder map keyed by the cell a player
    lands on.
    """

    def __init__(self, size: int, snakes: Sequence[Entity] = (), ladders: Sequence[Entity] = ()):
        self.size = size
        self._entities: Dict[int, int] = {}

        for s in snakes:
            if s.start <= s.end:
                raise BoardConfigError(f"snake start {s.start} must be greater than end {s.end}")
            self._add_entity(s.start, s.end)
        for l in ladders:
            if l.start >= l.end:
                raise BoardConfigError(f"ladder start {l.start} must be less than end {l.end}")
            self._add_entity(l.start, l.end)

    def _add_entity(self, start: int, end: int) -> None:
        if start <= 1 or start >= self.size:
            raise BoardConfigError(f"entity start {start} must be within (1, {self.size})")
        if start in self._entities:
            raise BoardConfigError(f"cell {start} already has a snake or ladder")
        self._entities[start] = end

    def resolve(self, cell: int) -> int:
        """Applies any snake/ladder at the landed-on cell, returning the
        final resting cell.
        """
        return self._entities.get(cell, cell)


class Player:
    def __init__(self, name: str):
        self.name = name
        self.position = 0


class NotEnoughPlayersError(Exception):
    pass


class GameAlreadyOverError(Exception):
    pass


class TooManyTurnsError(Exception):
    pass


class Game:
    """Drives turn order, dice rolls, and win detection for a single match."""

    def __init__(self, board: Board, dice: Dice, player_names: Sequence[str]):
        if len(player_names) < 2:
            raise NotEnoughPlayersError("need at least two players")
        self.board = board
        self.dice = dice
        self.players: List[Player] = [Player(name) for name in player_names]
        self._turn = 0
        self.winner: Optional[Player] = None

    def play_turn(self) -> Player:
        """Rolls the dice for the current player, applies the
        exact-landing rule and any snake/ladder at the destination, then
        rotates turn order to the next player. Returns the player who just
        moved.
        """
        if self.winner is not None:
            raise GameAlreadyOverError("game already has a winner")

        player = self.players[self._turn]
        roll = self.dice.roll()
        target = player.position + roll

        # Overshooting the last cell is not a legal move: the player stays put.
        if target <= self.board.size:
            player.position = self.board.resolve(target)

        if player.position == self.board.size:
            self.winner = player
        else:
            self._turn = (self._turn + 1) % len(self.players)
        return player

    @property
    def current_player(self) -> Player:
        """Returns whose turn it is next."""
        return self.players[self._turn]

    def play(self, max_turns: int) -> Player:
        """Runs turns until a winner emerges, guarding against runaway loops."""
        for _ in range(max_turns):
            self.play_turn()
            if self.winner is not None:
                return self.winner
        raise TooManyTurnsError(f"no winner after {max_turns} turns")


def _demo() -> None:
    board = Board(
        100,
        snakes=[Entity(99, 41), Entity(70, 34), Entity(52, 29)],
        ladders=[Entity(4, 25), Entity(21, 82), Entity(50, 95)],
    )
    game = Game(board, StandardDice(seed=42), ["Alice", "Bob"])
    winner = game.play(1000)
    print(f"{winner.name} wins at position {winner.position}!")


if __name__ == "__main__":
    _demo()
